amf-api-contract/shared/src/main/scala/amf/apicontract/internal/validation/shacl/graphql/GraphQLValidator.scala

```diff
@@ -6,12 +6,142 @@ import amf.core.internal.metamodel.Field
+              validateArgumentTypes(required, actual) ++ validateCovariance(required, actual)
+            case (required: GraphQLProperty, actual: GraphQLProperty) =>
+              validateCovariance(required, actual)
+            case (required: GraphQLOperation, _: GraphQLProperty) =>
+              validationInfo(
+                NodeShapeModel.Properties,
+                s"Field '${requiredField.name}' is missing it's required arguments: ${required.parameters.map(_.name).mkString(", ")}",
+                obj.annotations
+              )
+            case (required: GraphQLProperty, _: GraphQLOperation) =>
+              validationInfo(
+                NodeShapeModel.Properties,
+                s"Required field '${required.name}' has no arguments defined in '${requiredFieldClass.interface}' interface",
+                obj.annotations
+              )
+            case _ => None
+          }
+        case None =>
+          validationInfo(
+            NodeShapeModel.Properties,
+            s"Field '${requiredField.name}' from interface '${requiredFieldClass.interface}' is required in '${obj.name}'",
+            obj.annotations
+          )
+      }
+    }
+  }
+
+  private def validateArgumentTypes(required: GraphQLOperation, actual: GraphQLOperation): Seq[ValidationInfo] = {
+    val tuples = required.parameters.map { requiredArg =>
+      (requiredArg, actual.parameters.find(_.name == requiredArg.name))
+    }
+
+    val incorrectDatatypeValidations = tuples flatMap {
+      case (requiredArg: GraphQLParameter, None) =>
+        validationInfo(
+          AbstractParameterModel.Required,
+          s"Field '${actual.name}' is missing its required argument '${requiredArg.name}''",
+          actual.annotations
+        )
+      case (requiredArg: GraphQLParameter, Some(actualArg: GraphQLParameter)) =>
+        if (requiredArg.datatype != actualArg.datatype) {
+          val message = requiredArg.datatype match {
+            case Some(datatype) =>
+              s"Argument '${requiredArg.name}' of field '${required.name}' must be of type '$datatype''"
+            case None =>
+              s"Argument '${requiredArg.name}' of field '${required.name}' has an incorrect datatype"
+          }
+          validationInfo(AbstractParameterModel.Schema, message, actualArg.annotations)
+        } else None
+      case _ => None
+    }
+
+    val notDefinedValidations = actual.parameters
+      .filter { arg => !required.parameters.exists(_.name == arg.name) && arg.required }
+      .flatMap { arg =>
+        validationInfo(
+          AbstractParameterModel.Schema,
+          s"Required field '${required.name}' does not define a non-optional argument '${arg.name}'",
+          arg.annotations
+        )
+      }
+
+    incorrectDatatypeValidations ++ notDefinedValidations
+  }
+
+  private def validateCovariance(required: GraphQLField, actual: GraphQLField): Seq[ValidationInfo] = {
```

Comment:
Consider passing the RequiredField class so you can use the interface name in error messages